src/domain/flow-system/domain/src/entities/shared/flow_type.rs

```diff
@@ -52,7 +63,8 @@ impl DatasetFlowType {
                 flow_configuration_type == std::any::type_name::<TransformRule>()
             }
             DatasetFlowType::HardCompaction => {
-                flow_configuration_type == std::any::type_name::<CompactionRule>()
+                flow_configuration_type == std::any::type_name::<CompactionRuleMetadataOnly>()
+                    || flow_configuration_type == std::any::type_name::<CompactionRuleFull>()
```

Comment:
These string-based type compatibility checks look horrible...

Why can't we have something like this instead:

```rust
FlowConfiguration::check_compatible(&self, flow_type: DatasetFlowType) -> Result<(), IncompatibleFlowConfigError>;
```